### PR TITLE
Daemon images search refactoring

### PR DESCRIPTION
**Pull request by vdemeester**
Add Unit test to `daemon.SearchRegistryForImages`…
… and refactor a little bit some daemon on the way 🐙.
- Move `SearchRegistryForImages` to a new file (`daemon/search.go`) as `daemon.go` is getting pretty big.
- `registry.Service` is now an interface (allowing us to decouple it a little bit and thus unit test easily).
- Add some unit test for `SearchRegistryForImages`.
- Use `UniqueExactMatch` for search filters
- And use empty restore id for now in `client.ContainerStart`.

/cc @thaJeztah @runcom @calavera @stevvooe @aaronlehmann 
/cc @boucher @crosbymichael for checkpoint stuff.

🐸

### CODE REVIEW

**Comment by boucher** [comment]
Are there changes to the checkpoint stuff, or is it just pulling in the commit from engine-api?

**Comment by aaronlehmann** [comment]
> Are there changes to the checkpoint stuff, or is it just pulling in the commit from engine-api?

I gather that it's related to upstream changes in engine-api, but I also found this confusing. I think we need to do a better job of keeping engine in sync with engine-api and not syncing up in unrelated PRs.

**Comment by vdemeester** [comment]
@boucher @aaronlehmann yeah it's just the client signature that has been merged in engine-api before the changes I wanted to vendor :angel:.

@aaronlehmann do you suggest to do a PR only for vendoring ? (I thought about it but… :sweat_smile: )

**Comment by aaronlehmann** [comment]
> @aaronlehmann do you suggest to do a PR only for vendoring ? (I thought about it but… 😅 )

I wouldn't bother right now, but I think it's a good idea in general. Ideally the person who makes breaking changes to engine-api would open PRs for docker, swarm, etc. to keep those projects up to date.

**Comment by vdemeester** [comment]
@aaronlehmann yeah I do agree, we need more practice on this :angel:

File: api/client/start.go [hunk 1]
```diff
@@ -112,7 +112,7 @@ func (cli *DockerCli) CmdStart(args ...string) error {
 		})
 
 		// 3. Start the container.
-		if err := cli.client.ContainerStart(context.Background(), container); err != nil {
+		if err := cli.client.ContainerStart(context.Background(), container, ""); err != nil {
```
Comment (stevvooe) [inline api/client/start.go]: Allocate the context at the top of the function. It is useless if it is not shared (see above: a cancel gets completely ignored here.

File: registry/service.go [hunk 2]
```diff
@@ -13,29 +13,41 @@ import (
 	registrytypes "github.com/docker/engine-api/types/registry"
 )
 
-// Service is a registry service. It tracks configuration data such as a list
+// Service is the interface defining what a registry service should implement.
+type Service interface {
+	Auth(authConfig *types.AuthConfig, userAgent string) (status, token string, err error)
```
Comment (stevvooe) [inline registry/service.go]: While we're in here, let's consider plumbing the context.

File: api/client/start.go (hunk 1 above)
Comment (vdemeester) [inline api/client/start.go]: @stevvooe agree but I wonder if it would be better in another PR (:angel:).

Comment (stevvooe) [inline api/client/start.go]: Either way, but this code is just wrong.

Do we need some sort of best practices written up on context? Looking at the code and other PRs, I feel like there is some misunderstanding in effectively and using and propagating context.

Comment (vdemeester) [inline api/client/start.go]: Probably not a written best practice on context because it's available elsewhere, but yeah we nedd to fix the current code (mainly in `/api/client`. I'll work on it today :wink:

Comment (vdemeester) [inline api/client/start.go]: @stevvooe see #22878 to fix these :wink:

File: registry/service.go (hunk 2 above)
Comment (vdemeester) [inline registry/service.go]: @stevvooe hum why not, but I wonder how I should do it (i.e. just defining in the interface/methods and not using it inside letting other ppl using it or going through the implementation too). I'll take a look :wink:

File: registry/service.go [hunk 3]
```diff
@@ -7,35 +7,50 @@ import (
...
+	ResolveIndex(name string) (*registrytypes.IndexInfo, error)
+	Search(ctx context.Context, term string, authConfig *types.AuthConfig, userAgent string, headers map[string][]string) (*registrytypes.SearchResults, error)
+	ServiceConfig() *registrytypes.ServiceConfig
+	TLSConfig(hostname string) (*tls.Config, error)
+}
+
+// DefaultService is a registry service. It tracks configuration data such as a list
 // of mirrors.
-type Service struct {
+type DefaultService struct {
```
Comment (aaronlehmann) [inline registry/service.go]: The name `DefaultService` suggests that these services are pluggable, but in reality this is the only implementation, and the purpose of the interface seems to be to allow mocking for tests.

I can't think of a better name right now though.

**Comment by aaronlehmann** [comment]
LGTM

File: registry/service.go (hunk 3 above)
Comment (vdemeester) [inline registry/service.go]: @aaronlehmann yeah I'm bad at naming things and I did not find another term than Default… :sweat_smile:

File: registry/service.go [hunk 4]
```diff
@@ -7,35 +7,50 @@ import (
...
 
-// Service is a registry service. It tracks configuration data such as a list
+// Service is the interface defining what a registry service should implement.
+type Service interface {
+	Auth(ctx context.Context, authConfig *types.AuthConfig, userAgent string) (status, token string, err error)
+	LookupPullEndpoints(hostname string) (endpoints []APIEndpoint, err error)
+	LookupPushEndpoints(hostname string) (endpoints []APIEndpoint, err error)
+	ResolveRepository(name reference.Named) (*RepositoryInfo, error)
+	ResolveIndex(name string) (*registrytypes.IndexInfo, error)
+	Search(ctx context.Context, term string, authConfig *types.AuthConfig, userAgent string, headers map[string][]string) (*registrytypes.SearchResults, error)
```
Comment (thaJeztah) [inline registry/service.go]: @aaronlehmann @stevvooe wondering; should "search" be part of the registry interface, or be in it's own interface? I seem to recall "registry" in itself didn't have search facilities (although Docker Hub does)

**Comment by cpuguy83** [comment]
Do we know about this flakey test?

```
10:13:11 ----------------------------------------------------------------------
10:13:11 FAIL: docker_cli_wait_test.go:57: DockerSuite.TestWaitNonBlockedExitRandom
10:13:11 
10:13:11 docker_cli_wait_test.go:62:
10:13:11     c.Assert(err, checker.IsNil) //Container should have stopped by now
10:13:11 ... value *errors.errorString = &errors.errorString{s:"condition \"\"true\" == \"false\"\" not true in time"} ("condition \"\"true\" == \"false\"\" not true in time")
10:13:11 
10:13:18 
10:13:18 ----------------------------------------------------------------------
```

**Comment by thaJeztah** [comment]
@cpuguy83 don't think I've seen it

**Comment by thaJeztah** [comment]
LGTM, all green. We can look at https://github.com/docker/docker/pull/22861#discussion_r64369846 in a follow up (it was like this before, so 👍 )
